### PR TITLE
Auth: Add OIDC identities to identity cache and extract identity provider groups

### PR DESCRIPTION
**Pull request by markylaing**
This PR adds a new server configuration key `oidc.groups_claim`. If set, the value of this config key is:
1. Added to the `scope` of the Authorization Code Flow (UI).
2. Sent to the client as an `X-LXD-OIDC-groups-claim` header, to be added to the scope for the Device Authorization Grant Flow (CLI).
3. The value of the claim represents the groups that the identity is a member of at the identity provider level. It is extracted from both identity tokens (UI) and access tokens (CLI) and is expected to be a string slice.
4. Identity provider groups are added to the request context.
5. Identity provider groups are forwarded to other cluster members in the `X-LXD-forwarded-identity-provider-groups` header. The value of this header is a JSON encoded string.
6. Forwarded identity provider groups are also added to the request context.

#12816 must be merged first.

### CODE REVIEW

**Comment by markylaing** [comment]
@tomponline this is ready for review now.

Also, do we prefer `oidc.groups_claim` or `oidc.groups.claim` for this config key?

**Comment by tomponline** [comment]
> oidc.groups.claim

I would say, as it allows us to use oidc.groups.foo later if needed

File: lxd/auth/oidc/oidc.go [hunk 1]
```diff
@@ -178,6 +179,37 @@ func (o *Verifier) authenticateIDToken(ctx context.Context, w http.ResponseWrite
 	return claims.Subject, nil
 }
 
+// getGroupsFromClaims attempts to get the configured groups claim from the token claims and warns if it is not present
+// or is not a valid type. The custom claims are an unmarshalled JSON object.
+func (o *Verifier) getGroupsFromClaims(customClaims map[string]any) []string {
```
Comment (tomponline) [inline lxd/auth/oidc/oidc.go:221]: why map[string]any?

File: lxd/daemon.go [hunk 2]
```diff
@@ -334,12 +334,30 @@ func (d *Daemon) Authenticate(w http.ResponseWriter, r *http.Request) (trusted b
...
 	if d.oidcVerifier != nil && d.oidcVerifier.IsRequest(r) {
-		userName, err := d.oidcVerifier.Auth(d.shutdownCtx, w, r)
+		result, err := d.oidcVerifier.Auth(d.shutdownCtx, w, r)
 		if err != nil {
 			return false, "", "", err
 		}
 
-		return true, userName, api.AuthenticationMethodOIDC, nil
+		_, err = d.identityCache.Get(api.AuthenticationMethodOIDC, result.Subject)
+		if err != nil {
```
Comment (tomponline) [inline lxd/daemon.go:344]: I dont think we should be considering all errors from the .Get call to mean "missing identity", instead we should be inspection the error for a NotFound status error.

File: lxd/daemon.go [hunk 3]
```diff
@@ -334,12 +334,30 @@ func (d *Daemon) Authenticate(w http.ResponseWriter, r *http.Request) (trusted b
 	}
 
 	if d.oidcVerifier != nil && d.oidcVerifier.IsRequest(r) {
-		userName, err := d.oidcVerifier.Auth(d.shutdownCtx, w, r)
+		result, err := d.oidcVerifier.Auth(d.shutdownCtx, w, r)
 		if err != nil {
 			return false, "", "", err
```
Comment (tomponline) [inline lxd/daemon.go:340]: could we be wrapping the error here with more context, "Failed OIDC authentication: %w" or something like that?

File: lxd/daemon.go [hunk 4]
```diff
@@ -357,6 +357,12 @@ func (d *Daemon) Authenticate(w http.ResponseWriter, r *http.Request) (trusted b
 			updateIdentityCache(d)
 		}
 
+		if len(result.IdentityProviderGroups) > 0 {
```
Comment (tomponline) [inline lxd/daemon.go:361]: There's several authentication related vars added to the request context in daemon.createCmd.
I wondered if we should be keeping all of that logic together, either by moving it into this function or by moving this out to createCmd. Let me know.

File: lxd/daemon.go [hunk 5]
```diff
@@ -523,6 +524,16 @@ func (d *Daemon) createCmd(restAPI *mux.Router, version string, c APIEndpoint) {
...
 				ctx = context.WithValue(ctx, request.CtxForwardedUsername, r.Header.Get(request.HeaderForwardedUsername))
 				ctx = context.WithValue(ctx, request.CtxForwardedProtocol, r.Header.Get(request.HeaderForwardedProtocol))
+				forwardedIdentityProviderGroupsJSON := r.Header.Get(request.HeaderForwardedIdentityProviderGroups)
+				if forwardedIdentityProviderGroupsJSON != "" {
+					var forwardedIdentityProviderGroups []string
+					err = json.Unmarshal([]byte(forwardedIdentityProviderGroupsJSON), &forwardedIdentityProviderGroups)
+					if err != nil {
+						logger.Error("Failed unmarshalling identity provider groups from forwarded request header", logger.Ctx{"error": err})
+					} else {
+						ctx = context.WithValue(ctx, request.CtxForwardedIdentityProviderGroups, forwardedIdentityProviderGroups)
```
Comment (tomponline) [inline lxd/daemon.go:534]: Now im even more certain we should b populating the request contet in createCmd only :)

File: lxd/cluster/connect.go [hunk 6]
```diff
@@ -160,22 +160,22 @@ func ConnectIfVolumeIsRemote(s *state.State, poolName string, projectName string
...
-			// Replace node list with instance's cluster member node (which might be local member).
-			nodes = []db.NodeInfo{instNode}
-		} else {
+		if remoteInstance == nil {
 			// Volume isn't exclusively attached to an instance. Use local cluster member.
 			return nil, nil
 		}
+
+		var instNode db.NodeInfo
+		err = s.DB.Cluster.Transaction(context.TODO(), func(ctx context.Context, tx *db.ClusterTx) error {
```
Comment (tomponline) [inline lxd/cluster/connect.go:169]: good opportunity to use the s.ShutdownCtx var instead of context.TODO()

File: test/suites/oidc.sh [hunk 7]
```diff
@@ -11,12 +11,14 @@ test_oidc() {
 
   BROWSER=curl lxc remote add --accept-certificate oidc "${LXD_ADDR}" --auth-type oidc
   [ "$(lxc info oidc: | grep ^auth_user_name | sed "s/.*: //g")" = "unknown" ]
+  [ "$(lxd sql global "SELECT identifier, name, auth_method, type FROM identities WHERE type = 5 AND identifier = 'unknown' AND auth_method = 2" | wc -l)" = 5 ]
```
Comment (tomponline) [inline test/suites/oidc.sh:14]: nice

**Review by tomponline — COMMENTED** [review]
A handful of nits/questions, overall looks great.

File: lxd/daemon.go (hunk 5 above)
Comment (markylaing) [reply lxd/daemon.go:534]: Yep I'll make the change.

File: lxd/daemon.go (hunk 3 above)
Comment (markylaing) [reply lxd/daemon.go]: I've made this change in a separate commit. There was a type assertion on this error for `err.(*oidc.AuthError)`. I've now made the `oidc.Verifier` return the error by value and am using `errors.As` to inspect it.

File: lxd/daemon.go [hunk 8]
```diff
@@ -482,7 +482,7 @@ func (d *Daemon) createCmd(restAPI *mux.Router, version string, c APIEndpoint) {
 			}
 
 			// If not a macaroon discharge request, return the error
-			_, ok = err.(*bakery.DischargeRequiredError)
+			_, ok := err.(*bakery.DischargeRequiredError)
```
Comment (markylaing) [inline lxd/daemon.go:485]: Looks like I missed some macaroon based authentication stuff in #12830. I'll send a follow up PR.

**Comment by markylaing** [comment]
> Needs an API extension too please

What would the extension be for? This doesn't add any functionality (yet). Or is it just because we're adding a new config key?

**Comment by tomponline** [comment]
> > Needs an API extension too please
> 
> What would the extension be for? This doesn't add any functionality (yet). Or is it just because we're adding a new config key?

Yep the new config key is new functiona

**Comment by markylaing** [comment]
> Yep the new config key is new functionality

🤔 I suppose it is. You can set some config.

I'll update the PR on Monday and will add an API extension check in the client for adding this config key.

File: lxd/daemon.go (hunk 8 above)
Comment (markylaing) [reply lxd/daemon.go:485]: Follow up PR here #12860

**Comment by markylaing** [comment]
> Needs an API extension too please

Added an `oidc_groups_claim` API extension.

**Comment by markylaing** [comment]
@tomponline I've made the requested changes. I haven't checked for the `oidc_groups_claim` API extension when attempting to set this config key in the client because there doesn't seem to be any existing API extension checks for config keys.

File: doc/api-extensions.md [hunk 9]
```diff
@@ -2322,3 +2322,9 @@ mounted.
 
 The API extension adds indication whether the LXD version is an LTS release.
 This is indicated when command `lxc version` is executed or when `/1.0` endpoint is queried.
+
+## `oidc_groups_claim`
```
Comment (tomponline) [inline doc/api-extensions.md:2326]: This should be in the previous commit, and in the format of other API extensions, .e.g. https://github.com/canonical/lxd/commit/841c023dad44872c89cd193e52ae3b98fa8c0334

**Review by tomponline — CHANGES_REQUESTED** [review]
just an api extension commit change

**Review by tomponline — APPROVED** [review]

**Pull request merged**